packages/language/src/language-server/connection-handler.ts

```diff
@@ -54,116 +63,183 @@ export function startLanguageServer(connection: Connection): void {
-                newText: params.newName,
-              }) satisfies TextEdit,
-          );
-        }
-      });
+  }
+
+  const offset = textDocument.offsetAt(params.position);
+  const definition = referencesRequest(compilationUnit, uri, offset);
 
-      return {
-        changes,
-      };
+  const lspDefinitions: Location[] = [];
+  for (const def of definition) {
+    let doc = TextDocuments.get(def.uri);
+
+    if (!doc) {
+      const content = FileSystemProviderInstance.readFileSync(
+        URI.parse(def.uri),
+      );
+      doc = TextDocument.create(def.uri, "pli", 0, content);
+      TextDocuments.set(doc);
     }
 
+    if (doc) {
+      const range = rangeToLSP(doc, def.range);
+      lspDefinitions.push({
+        uri: def.uri,
+        range,
+      });
+    }
+  }
+
+  return lspDefinitions;
+}
+
+export function handleSemanticTokensRequest(
+  params: SemanticTokensParams,
+  compilationUnitHandler: CompletionUnitHandler,
+): SemanticTokens {
+  const uri = params.textDocument.uri;
+  const textDocument = TextDocuments.get(uri);
+  const compilationUnit = compilationUnitHandler.getCompilationUnit(
+    URI.parse(uri),
+  );
+
+  if (!textDocument || !compilationUnit) {
+    return { data: [] };
+  }
+
+  return {
+    data: semanticTokens(textDocument, compilationUnit),
+  };
+}
+
+export function handleDocumentHighlightRequest(
+  params: TextDocumentPositionParams,
+  compilationUnitHandler: CompletionUnitHandler,
+): DocumentHighlight[] {
+  const uri = UriUtils.normalize(params.textDocument.uri);
+  const textDocument = TextDocuments.get(uri);
+  const parsedUri = URI.parse(uri);
+  const unit = compilationUnitHandler.getCompilationUnit(parsedUri);
+
+  if (!textDocument || !unit) {
+    return [];
+  }
+
+  const offset = textDocument.offsetAt(params.position);
+  const definitions = getReferenceLocations(unit, parsedUri, offset);
+  return definitions
+    .filter((e) => e.uri === uri)
+    .map((def) =>
+      DocumentHighlight.create(rangeToLSP(textDocument, def.range)),
+    );
+}
+
+export function handleRenameRequest(
+  params: {
+    textDocument: { uri: string };
+    position: { line: number; character: number };
+    newName: string;
+  },
+  compilationUnitHandler: CompletionUnitHandler,
+): { changes: { [uri: string]: TextEdit[] } } | null {
+  const uri = params.textDocument.uri;
+  const textDocument = TextDocuments.get(uri);
+  const parsedUri = URI.parse(uri);
+  const unit = compilationUnitHandler.getCompilationUnit(parsedUri);
+
+  if (!textDocument || !unit) {
     return null;
+  }
+
+  const offset = textDocument.offsetAt(params.position);
+  const renameLocations = renameRequest(unit, parsedUri, offset);
+  const changes = mapValues(renameLocations, (locations, key) => {
+    let doc = TextDocuments.get(key);
+
+    if (!doc) {
+      const content = FileSystemProviderInstance.readFileSync(URI.parse(key));
+      doc = TextDocument.create(key, "pli", 0, content);
+      TextDocuments.set(doc);
```

Comment:
Suggestion: We should have some kind of try-catch for this, as any file-system error will prevent the request from completing.